### PR TITLE
Add an asynchronous single GPU dataloader sexample

### PR DESCRIPTION
**Pull request by HenryJia**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?   
- [x] Did you write any new necessary tests?  
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CHANGELOG.md)?

## What does this PR do?
Address #1454 #1404 #1316

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pl_examples/utils/loaders.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    queue_size: Integer
+        Size of the queue used to store the data loaded to the device
+    """
+
+    def __init__(self, dataset, device, num_workers=8, queue_size=10, **kwargs):
+        super(AsynchronousLoader, self).__init__(dataset=dataset,
+                                                 pin_memory=True,
+                                                 num_workers=num_workers,
+                                                 **kwargs
+                                                 )
```
Comment (Borda) [inline pl_examples/utils/loaders.py:32]:
```suggestion
        super().__init__(dataset=dataset, pin_memory=True, num_workers=num_workers, **kwargs)
```
just to be consistent...

File: pl_examples/utils/loaders.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+            self.worker.join()
+            raise StopIteration
+        else:  # Otherwise return the next batch
+            out = self.queue.get()
+            self.queue.task_done()
+            self.idx += 1
+        return out
+
+    def __len__(self):
+        return super(AsynchronousLoader, self).__len__()
```
Comment (Borda) [inline pl_examples/utils/loaders.py:77]:
```suggestion
        return super().__len__()
```

File: pl_examples/utils/loaders.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        self.worker = Thread(target=self.load_loop)
+        self.worker.daemon = True
+        self.worker.start()
+        return self
+
+    def __next__(self):
+        # If we've reached the number of batches to return
+        # or the queue is empty and the worker is dead then exit
+        done = not self.worker.is_alive() and self.queue.empty()
+        done = done or self.idx >= super(AsynchronousLoader, self).__len__()
```
Comment (Borda) [inline pl_examples/utils/loaders.py:64]:
```suggestion
        done = done or self.idx >= super().__len__()
```

File: pl_examples/basic_examples/async_gpu_template.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
+"""
+Runs a model on a single node across multiple gpus.
```
Comment (Borda) [inline pl_examples/basic_examples/async_gpu_template.py:2]: Pls add some a bit longer description and a note changelog

File: pl_examples/utils/loaders.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+from queue import Queue
+
+import torch
+from torch import Tensor
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.utils.data import DataLoader
+
+
+class AsynchronousLoader(DataLoader):
```
Comment (justusschock) [inline pl_examples/utils/loaders.py:11]: One question on this: Let's say I use a custom data loader. Can I also use this async loader? I think I can't since the logic I implemented myself would be missing, right? Is there a possibility to just wrap the already specified loader instead of replacing it?

Comment (HenryJia) [reply pl_examples/utils/loaders.py:11]: Yes, it is be possible. The only issue I've seen so far arises when the custom data loader is infinite (or of an unknown length). Then the termination of the loading thread becomes more complicated

Comment (justusschock) [reply pl_examples/utils/loaders.py]: Let's just say, that the user can give an explicit length argument (defaults to None) and include a check, whether this argument was set or the loader contains a length

**Comment by veritas9872** [comment]
@HenryJia The __len__ attribute seems to be wrong. To the best of my knowledge, DataLoaders do not have __len__. To access the length, you should use DataLoader.Dataset to access the dataset of the data loader.

**Comment by justusschock** [comment]
@veritas9872 That's not correct. The pytorch data loader does have a length that computes the number of samples based on batch size and dataset/sampler length. It's the dataset that does not necessarily have a length.

File: pl_examples/utils/loaders.py [hunk 6]
```diff
@@ -8,28 +8,37 @@
...
+        Any additional arguments to pass to the dataloader if we're constructing one here
     """
 
-    def __init__(self, dataset, device, num_workers=8, queue_size=10, **kwargs):
-        super(AsynchronousLoader, self).__init__(dataset=dataset,
-                                                 pin_memory=True,
-                                                 num_workers=num_workers,
-                                                 **kwargs
-                                                 )
+    def __init__(self, dataset=None, dataloader=None, device=torch.device('cuda', 0), q_size=10, **kwargs):
```
Comment (justusschock) [inline pl_examples/utils/loaders.py:37]: One more question: Can we maybe change the order of dataset and data loader here? I think by default the user specifies a loader and just wants it to be wrapped. Therefore it should also be the first positional argument here.

Also I'm thinking if we can maybe check for types with a single argument here.
Something like 
```python
if is instance(data, torch.utils.data.Dataset):
    data = ... # Instatiate loader
# otherwise assume data is already a loader providing an `__iter__` and `__next__` interface
```

File: pl_examples/utils/loaders.py (hunk 1 above)
Comment (HenryJia) [reply pl_examples/utils/loaders.py]: Already dealt with this

File: pl_examples/basic_examples/async_gpu_template.py (hunk 4 above)
Comment (HenryJia) [reply pl_examples/basic_examples/async_gpu_template.py]: Done!

File: pl_examples/utils/loaders.py (hunk 6 above)
Comment (HenryJia) [reply pl_examples/utils/loaders.py:37]: Will do, make sense

File: pl_examples/utils/loaders.py (hunk 2 above)
Comment (HenryJia) [reply pl_examples/utils/loaders.py]: Yep, done!

File: pl_examples/utils/loaders.py (hunk 6 above)
Comment (HenryJia) [reply pl_examples/utils/loaders.py]: @justusschock done!

File: pl_examples/basic_examples/async_gpu_template.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+import torchvision.transforms as transforms
+from torchvision.datasets import MNIST
+
+import pytorch_lightning as pl
+from pytorch_lightning import _logger as log
+
+from pl_examples.models.lightning_template import LightningTemplateModel
+from pl_examples.utils.loaders import AsynchronousLoader
+
+SEED = 2334
```
Comment (Borda) [inline pl_examples/basic_examples/async_gpu_template.py:19]: this shall be fixed in #1572

**Comment by HenryJia** [comment]
@justusschock I thought about adding it in to lightning itself instead of just as an example, but I'm not sure where exactly to stick it. It's not very generalisable as well as it's only single GPU compatible. To get it to work with infinite dataloaders would add a bit of complexity as I'd need to add in thread stopping conditions. I figured sticking with the PyTorch/PyTorch lightning philosophy of modularity, I'd keep it as an wrapper example with dataloader

File: pl_examples/utils/loaders.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
+from threading import Thread
+from queue import Queue
+
+import torch
+from torch import Tensor
+import torch.nn as nn
+import torch.nn.functional as F
```
Comment (awaelchli) [inline pl_examples/utils/loaders.py:7]: It seems some of these imports are unused

File: pl_examples/basic_examples/async_gpu_template.py (hunk 7 above)
Comment (awaelchli) [reply pl_examples/basic_examples/async_gpu_template.py:19]: Is the seed actually needed? The example does not actually produce any resutlts, it's just a demo for a dataloader, right?

File: pl_examples/utils/loaders.py (hunk 8 above)
Comment (HenryJia) [reply pl_examples/utils/loaders.py:7]: Yep, will deal with those

Comment (HenryJia) [reply pl_examples/utils/loaders.py]: @awaelchli Done! Although I think it's failing the CI build and tests because I resolved the conflict with a merge from master and master is currently failing

**Comment by williamFalcon** [comment]
@HenryJia this is awesome, but i think this belongs in bolts? 
@PyTorchLightning/core-contributors thoughts?

**Comment by awaelchli** [comment]
Could be a good place for these kind of examples

**Comment by awaelchli** [comment]
maybe could go to bolts.dataloaders?

**Comment by HenryJia** [comment]
Following some discussions with William, we agreed that this would be best added to the datamodules section of bolts instead. As such, I'm closing this PR for now unless anyone else has something to object about it. I will open a new PR with this for bolts soon once I get around my final year university exams.
